Allow setting pooling allocation in the C API

fridaymore opened the pull request:
We are troubleshooting performance / scaling issues in our project that uses Wasmtime and Pooling Allocation was recommended. Currently, we are not able to enable Pooling Allocation through the C API. 


Note: mpk options are not exposed.

alexcrichton reviewed (commented):
Thanks! Would you be up for refactoring this to use an object-style API along the lines of `wasm_config_t` where a new `wasmtime_pooling_config_t` object was added and the properties here are functions to configure each field? That way it's possible to configure just a single option while leaving all the others at their defaults, while currently it would require configuring all options at the same time.

Additionally as a minor thing, mind putting `#ifdef` guards around these definitions for the pooling allocator cfg being turned on?

fridaymore commented:
> Thanks! Would you be up for refactoring this to use an object-style API along the lines of `wasm_config_t` where a new `wasmtime_pooling_config_t` object was added and the properties here are functions to configure each field? That way it's possible to configure just a single option while leaving all the others at their defaults, while currently it would require configuring all options at the same time.
> 
> Additionally as a minor thing, mind putting `#ifdef` guards around these definitions for the pooling allocator cfg being turned on?

Thanks - I switched to an object-style API and added the `#ifdef` guards for feature gating.  Can you opine on memory management for `wasmtime_pooling_config_t`? I am assuming I need to provide a way to free the memory that was allocated on the heap?

alexcrichton commented:
Yeah adding a free method for it I think is the way to go. Otherwise "cloning out" the configuration makes sense to me as well.

Mind also adding `wasmtime_*` prefixes to the methods/types here too?

alexcrichton commented on crates/c-api/include/wasmtime/config.h:
```diff
@@ -542,6 +542,206 @@ wasmtime_config_host_memory_creator_set(wasm_config_t *,
 
+#ifdef WASMTIME_FEATURE_POOLING_ALLOCATOR
+
+typedef struct wasmtime_pooling_allocation_config_t
+    wasmtime_pooling_allocation_config_t;
+
+WASM_API_EXTERN wasmtime_pooling_allocation_config_t *
+wasmtime_pooling_allocation_config_new();
+
+WASM_API_EXTERN void wasmtime_pooling_allocation_config_delete(
+    wasmtime_pooling_allocation_config_t *);
```
Can you add some documentation to these functions/types? Just some basic docs is fine, but for example indicating that `*_new()` must be deallocated with `*_delete()` and perhaps linking to the Rust documentation for `wasmtime_pooling_allocation_config_t`.

Also for `wasmtime_pooling_allocation_strategy_set` down below can you document that it doesn't take ownership of the pooling config and the pooling config still needs to be manually deleted?

fridaymore replied:
Done

fridaymore commented on crates/c-api/include/wasmtime/config.h:
```diff
@@ -542,6 +542,206 @@ wasmtime_config_host_memory_creator_set(wasm_config_t *,
+WASMTIME_POOLING_ALLOCATION_CONFIG_PROP(total_gc_heaps, uint32_t)
+
+WASM_API_EXTERN void wasmtime_pooling_allocation_strategy_set(
+    wasm_config_t *, wasmtime_pooling_allocation_config_t *);
```
Done

fridaymore commented on crates/c-api/include/wasmtime/config.h:
```diff
@@ -542,6 +542,206 @@ wasmtime_config_host_memory_creator_set(wasm_config_t *,
+WASM_API_EXTERN void wasmtime_pooling_allocation_strategy_set(
+    wasm_config_t *, wasmtime_pooling_allocation_config_t *);
+
+#endif
```
Done
